Introduce causal mask

mikekgfb opened the pull request:
Summary: Introduce causal mask

This PR introduces a causal mask option is_causal, since current custom kernels do not support arbitrary masks.
(Also, building these masks is expensive, and avoiding to build them is a significant opportunity since causal mask is extremely common for training Transformer models.)

Test Plan: sandcastle & github ci/cd

Differential Revision: D41723137

danthe3rd commented on torch/nn/modules/transformer.py:
```diff
@@ -278,8 +278,21 @@ def forward(self, src: Tensor, mask: Optional[Tensor] = None, src_key_padding_ma
+            if _causal_mask:
+                raise RuntimeError("specify either mask or _causal_mask, but not both")
+
+            sz = mask.size(0)
+            causal_comparison = torch.triu(
+                torch.ones(sz, sz, device=mask.device) * float('-inf'), diagonal=1
+            ).to(mask.dtype)
```
this might be very expensive, and will run every time the user provides a mask. It will also use quite a lot of memory, potentially negating the benefits from using memory-efficient kernels

mikekgfb replied:
I think the preferred way will be to surface this as the causal mask flag as part of the API.  I am/was hoping that we might rely on torch.compile() to constant fold it.  (And maybe, if it is unable to do so, we might have it optimized away by torch.compile() ?

albanD commented on torch/nn/functional.py:
```diff
@@ -4922,6 +4922,7 @@ def multi_head_attention_forward(
     key_padding_mask: Optional[Tensor] = None,
     need_weights: bool = True,
     attn_mask: Optional[Tensor] = None,
+    _causal_mask: bool = False
```
Just mentioning that doesn't really make the arguments private. It will still show up in the signature of the function on the doc page.
This function is not actually documented on the website so the point is moot-ish but you can see where it is mentioned in https://docs-preview.pytorch.org/90508/jit_builtin_functions.html#supported-pytorch-functions

mikekgfb replied:
TBH, I don't think it should be private -- mirroring the concerns here from dan3, I think it will be preferable to give users the explicit option.

albanD commented on torch/nn/modules/transformer.py:
```diff
@@ -278,8 +283,26 @@ def forward(self, src: Tensor, mask: Optional[Tensor] = None, src_key_padding_ma
+            if is_causal:
+                raise RuntimeError("specify either mask or is_causal, but not both")
+
+            # sz = mask.size(0)
```
Should this be removed?

mikekgfb replied:
removed.

mikekgfb replied:
> Change sounds ok. Docstrings need to be updated.

Docstring updated.

albanD reviewed (commented):
Change sounds ok. Docstrings need to be updated.

mikekgfb commented on torch/nn/modules/transformer.py:
```diff
@@ -623,8 +643,16 @@ def __setstate__(self, state):
+        memory_mask: Optional[Tensor] = None,
+        tgt_key_padding_mask: Optional[Tensor] = None,
+        memory_key_padding_mask: Optional[Tensor] = None,
+        is_causal: bool = False,
```
added

mikekgfb commented on torch/nn/functional.py:
```diff
@@ -4929,6 +4929,7 @@ def multi_head_attention_forward(
     static_k: Optional[Tensor] = None,
     static_v: Optional[Tensor] = None,
     average_attn_weights: bool = True,
+    is_causal: bool = False,
```
added

albanD reviewed (approved):
The change itself sounds ok.
You most likely want to rebase above master to avoid trunk failures.

Commit:
Introduce causal mask (#90508)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/90508

Introduce causal mask

Test Plan: sandcastle & github ci/cd

Reviewed By: albanD, mleshen

Differential Revision: D41723137

fbshipit-source-id: 5cb222d672e8f174a1447a91414068ba87490dc3

mikekgfb commented:
@pytorchbot -f merge

mikekgfb commented:
@pytorchbot merge -f

mikekgfb commented:
@pytorchbot merge -f "lint runner was broken, as per Eli. Using -f as suggested by albanD"
